### ISSUE TITLE
Review polyfill approach to writing progressively-enhanced, x-browser Javascript

### ISSUE THREAD

**Issue by adamsilver** [alphagov/govuk-frontend#676]
I heard the project is considering a polyfill approach to JavaScript cross-browser support.

I spoke to Nick Colley on Slack about the [disadvantages of polyfills](https://adamsilver.io/articles/the-disadvantages-of-javascript-polyfills/) and how we might want to consider a more robust approach moving forwards.

The approach is discussed in (my article) [Progressively Enhanced JavaScript](https://adamsilver.io/articles/progressively-enhanced-javascript/), but I learnt this from Peter Michaux’s [Cross Browser Widgets](http://peter.michaux.ca/articles/cross-browser-widgets) article (don’t let the date put you off; the best techniques usually stand the test of time).

I’m happy to discuss/help further with this, but just getting this down as an issue for now.

**Comment by nickcolley** [comment]
We're currently in a halfway place between this approach and a strict polyfilling approach.

You can see the details component uses the approach described here: https://github.com/alphagov/govuk-frontend/blob/master/src/details/details.js#L11

Whereas the Checkbox component uses the (if we continue) the new approach:
https://github.com/alphagov/govuk-frontend/blob/master/src/checkboxes/checkboxes.js

**Comment by adamsilver** [comment]
It seems you've got a lot of the code necessary written to create a wrapper interface as opposed to polyfills. So with a little bit of work, you could avoid the polyfill problem.

**Comment by nickcolley** [comment]
1. They augmenting host objects
> This is a valid point but seems to be being made mainly on the basis of extending functionality that is not part of the stable ECMAScript specification. Which is not what we're doing.

2. Feature detection is not enough
> Well reasoned and makes sense, however the detections used by the Financial Times account for this. Where necessary they do more than check the prototype of a global feature.

3. They intertwine browser and application logic
> This argument seems to be made without considering polyfills, I agree that it's a bad idea to expose your application logic to complexities with different browser but both utilities and polyfilling achieves the same goal here?

4. We rarely need need the full API
> This makes sense.

5. They come with caveats
> This is definitely a risk.

---

Another risk is that we cannot be sure our polyfills do not break third party code.

I think the biggest positive with polyfilling is we are writing JavaScript that does not require developers to understand our abstractions. If you understand `addEventListener` it'll just work.

**Comment by adamsilver** [comment]
Thanks @nickcolley 

> This is a valid point but seems to be being made mainly on the basis of extending functionality that is not part of the stable ECMAScript specification. Which is not what we're doing.

What do you mean by stable here? Every host object API is subject to the pitfalls of augmentation.

> the detections used by the Financial Times account for this. Where necessary they do more than check the prototype of a global feature.

Agreed—I should amend my article. It's not a fault of polyfills per se. Thanks

> This argument seems to be made without considering polyfills, I agree that it's a bad idea to expose your application logic to complexities with different browser but both utilities and polyfilling achieves the same goal here?

Agreed.

**Comment by nickcolley** [comment]
> What do you mean by stable here? Every host object API is subject to the pitfalls of augmentation.

My interpretation is that the biggest issue with modifying native APIs is when it is done with no regard to standards. For example when MooTools used 'flatten' (https://github.com/tc39/proposal-flatMap/pull/56).

When polyfilling stable and accepted standards, this risk is not there? (With point five around cavets still in mind)

**Comment by adamsilver** [comment]
> My interpretation is that the biggest issue with modifying native APIs is when it is done with no regard to standards. For example when MooTools used 'flatten' (tc39/proposal-flatMap#56).

That's not the main problem and it's important to note the differences between native built-ins and host objects. 

Host objects are by far and away the most dangerous. So `el.addEventListener` is far more dangerous than say `Array.prototype.filter`.

Your point speaks to this issue:

- http://perfectionkills.com/whats-wrong-with-extending-the-dom/#chance_of_collisions

But there are some bigger issues:

- http://perfectionkills.com/whats-wrong-with-extending-the-dom/#lack_of_specification
- http://perfectionkills.com/whats-wrong-with-extending-the-dom/#host_objects_have_no_rules
- http://perfectionkills.com/whats-wrong-with-extending-the-dom/#performance_overhead
- http://perfectionkills.com/whats-wrong-with-extending-the-dom/#ie_dom_is_a_mess (admittedly, the browser is probably not on our radar anymore, but see the point about host objects having no rules)

**Comment by adamsilver** [comment]
Extension and polyfills are exactly the same thing here.

You take a browser that doesn't have an API, let's say `el.addEventListener`. Then you extend `el` (or objects of that type such as HTMLElement) with a user-defined `addEventListener` method.

**Comment by nickcolley** [comment]
I've spent some time reading the linked articles in detail.

There doesn't seem to be good evidence that these arguments apply to polyfill approaches, and seem to be reflecting on Prototype/MooTools abuse of native APIs.

The author's [follow up article](http://perfectionkills.com/extending-native-builtins/) includes arguments that seem to be in favour of polyfilling, with only exception being 'host objects'.

Given this the current approach does not seem as risky as suggested, but given GOV.UK Frontend is expected to be used in places we cannot anticipate maybe we should err on the side of caution.

This leaves us a few options:

1. Utility approach only
Pros:
- No chance of conflicts with third party scripts. 
Cons:
- A GOV.UK Frontend abstraction that developers need to learn.
- We would have to author these ourselves (or contributors).
- Not using JavaScript APIs, may make techniques such as `bind` cumbersome to author. 

(Note: One argument could be to continue using jQuery, but 1.x is no longer maintained and newer versions do not have the browser support we require)

2. Continue with Polyfills
Pros:
- Very easy to remove in the future
- JavaScript written is standard and transferable.
Cons:
- Possibility of conflicts or bugs when GOV.UK Frontend is loaded with JavaScript we can't test with.

3. Only polyfill ES5 features, using utilities instead of DOM polyfills.
This is a compromise, but would satisfy the issues, so could be an option.

Interested to hear what the rest of the community / team thinks. Thanks @adamsilver for bearing with me 😄

**Comment by frankieroberto** [comment]
I'm quite pro polyfills, mainly on the basis that they help developers to understand the native API rather than having to learn an abstraction. This is especially important with organisations such as government which have highly aggregated and unstable teams (eg with lots of contractors coming and going, and employees moving around lots).

The ability to easily drop polyfills in future is another bonus.

That said, I think it does put an onus on the govuk team to promote the 'best' way to do feature detection for each individual feature, that works across browsers and with both the native and polyfilled implementations.

There are also some features (such as the Drag and Drop API) which are impossible to feature-detect accurately without browser sniffing (blame mobile Safari, see https://stackoverflow.com/questions/2856262/detecting-html5-drag-and-drop-support-in-javascript).

**Comment by adamsilver** [comment]
I think we need to be careful about convenience for devs vs stability and resilience for users.

Although, I do think that polyfills are actually a burden on developers too.

I agree that documentation would be needed, but not sure that's a bad thing. Do the hard work to make it simple. :)

**Comment by nickcolley** [comment]
Update: I have proposed a piece of work to spike this proposal and see if it's possible, which we'll do within our current sprint. 

https://trello.com/c/Bef9APKQ/959-spike-swapping-polyfill-approach-with-utilities

**Comment by adamsilver** [comment]
@pauldwaite I've not yet had the chance to fully present what the alternative to polyfills would be. I apologise for that as the onus is on me.

If I had, I'd hope to show you that actually, a library is less work and can just as easily be ripped out in the future too.

Bare in mind that with polyfills, if you rip them out, then the interface is likely to _break _; so you'd have to be able to either:

- guarantee that _no_ users _ever_ will use a browser that lacks support for said method/property/feature; or
- simply not care that one or more users experienced that breakage.

These are two scenarios that as devs/designers we wouldn't have to ever think about again, given what I'm proposing.

**Comment by pauldwaite** [comment]
@adamsilver I get your point about the consequences of removing polyfills. I think we already don’t care than one or more users experience breakage: the service I’m working on will definitely break in IE 6 and 7, and I think we’re fine with that.

**Comment by frankieroberto** [comment]
@adamsilver providing that javascript is only ever progressive enhancement, and that feature-detection is used, removing the polyfill doesn’t break things, it just means that those users get the non-javascript experience.

**Comment by adamsilver** [comment]
Polyfills extend native built-ins and host objects. The main problems are with host objects.

For some examples of problems read the entire section here:

http://perfectionkills.com/whats-wrong-with-extending-the-dom/#host_objects_have_no_rules

That's a good start to understand the problem with messing with host objects.

**Comment by frankieroberto** [comment]
@adamsilver I don't think there are _any_ available polyfills for `querySelectorAll` (especially given that the CSS selector syntax is so large and ever-expanding), so I suspect that for that API, the only question is about how to feature-detect it (so that older browsers don't throw a javascript error)?

Which means either:

```javascript
if ('querySelectorAll' in navigator) {
   var links = document.querySelectorAll('main a');
}
```

or an abstraction like:

```javascript

if ('querySelectorAll' in navigator) { 
   function query(selector) {
      return document.querySelectorAll(selector);
   }
}

if (query) { 
   var links = query('main a'); }
}
```

The former has the advantage (in my view) of exposing developers to the actual native API method name, whereas the latter has the advantage of developers not having to think about the best way to do the feature detection (or forgetting to do it).

**Comment by adamsilver** [comment]
Thanks @frankieroberto 

It's `querySelectorAllin window` (or better yet use `isHostMethod()`) :D

Question: should all developers/designers who are choosing to use / relying on a library have/want/need to learn the native/host APIs?

**Comment by pauldwaite** [comment]
> Question: should all developers/designers who are choosing to use / relying on a library have/want/need to learn the native/host APIs?

The native/host APIs are always there. Libraries on top of them come and go.

**Comment by frankieroberto** [comment]
@adamsilver the problem with libraries are that there's so many of them! The chances of every government department agreeing to use the same ones seems slim, let alone the wider sector (and it's worth bearing in mind that many of the designers/developers working on government services are contractors or on secondment).

Learning the native API is always useful, in my view.  We learnt that the hard way with jQuery... ;-)

I suspect the bigger issue though will be that, as APIs like `addEventListener` become increasingly known and used (especially by those who don't have to worry so much by older browsers), it’s much easier for developers not to have to learn a library.

That said, I acknowledge the downside of having to teach/show people the best ways to do feature detection.

**Comment by pauldwaite** [comment]
In terms of feature detection, there is also the BBC’s “cutting the mustard” approach. (I like to think this is named after an obscure 70s sitcom.)

As opposed to putting a (fairly thin) library on top of the native APIs (as I think @adamsilver is suggesting), they essentially do one-liner feature detection for a few key features at the start of their script, and give up if any of the features aren’t present.

It’s a bit blunt, and probably easy to miss when reading existing code, compared to the library approach.

**Comment by adamsilver** [comment]
The solution I'm poorly trying to communicate is that we could end up writing JS a bit like this:

```
if(govuk.hasFeatures('addEventListener', 'querySelectorAll', 'isArray')) {
	
	// define a component that uses these methods such as a tab component

	function Tab(blah) {
		var tabs = govuk.querySelectorAll('.blah');
		govuk.addEventListener(tabs, 'click', ...);
	}

	Tab.prototype.blah = function() {

	}

}```

This is just an example API.

**Comment by leekowalkowski-hmrc** [comment]
They are confusing reasons against polyfills.  

I didn't understand the how the [approach described for document.querySelectorAll](https://github.com/alphagov/govuk-frontend/issues/676#issuecomment-386222493) was actually useful?  So instead of using `querySelectorAll`, I call a new function called `query`, which doesn't have a solution within it because it relies on `querySelectorAll`?  I feel like I missed the point.  I certainly wouldn't use the alternative `query` because it's not apparent what the benefit is.  It appears to be a needless wrapper.

With `attachEvent`, if it does not do event delegation, then polyfill or otherwise, you cannot use event delegation. No?  So I didn't understand how event delegation itself is a concern that influences whether or not you should polyfill.

I can only make sense of this as "there are some situations where polyfilling is a no-go, therefore, never use polyfills for anything".  Which feels to me like throwing the baby out with the bath water.

My approach to adopting semi-supported technology on public websites is that if it cannot easily be polyfilled (e.g. if one particular browser makes it impossible: 'Trying to overwrite "target" property of event object in Mozilla throws TypeError'), testing should be sufficient to discover this is the case, and should result in the decision to avoid the semi-supported technology.  It's often a nice-to-have and not driven by the users' needs, anyway.

This does mean I'm often working with feature set which of the lowest common denominator available natively, incidentally.  Which means I don't create polyfills lightly.  My commercial sites only have `Array.from` and `Array.indexOf` as polyfills.  I tend to use the ancient `element.onclick` style event handlers because I can design away any conflicts since I don't have the misfortune of dependencies on 3rd party libraries - modern native capability is good enough for everything.

I can appreciate the arguments against polyfills that are too ambitious or if the target audience is everybody, everywhere - the polyfill may fail in unexpected ways, interfere with other things, or not satisfy a niche case.  I think this comes from the desperation to use a new feature for the sake of it, rather than the fact we chose to polyfill it.

For example, writing a custom 'reveal password' feature, polyfill or not, is doomed to be surpassed by native implementations over time irrespective how good a job was made of it.

**Comment by pauldwaite** [comment]
I can definitely see the benefit of a feature-detection library (like `govuk.hasFeatures`).

What I’m a bit leery about is a library that provides an API on top of those features (`govuk.querySelectorAll`).

**Comment by adamsilver** [comment]
> I’d be interested in the “less work” comparison. I’ve mainly grabbed already-written polyfills, mostly from MDN. The work has been pretty minimal, if you ignore the other people’s work in writing the polyfills in the first place.

If you ignore the work of writing the polyfill, then you ignore the work of writing the alternative to polyfills.

**Comment by adamsilver** [comment]
> I didn't understand the how the approach described for document.querySelectorAll was actually useful? So instead of using querySelectorAll, I call a new function called query, which doesn't have a solution within it because it relies on querySelectorAll? I feel like I missed the point.

You either missed the point, or I failed to explain (probably the latter).

With the wrapper code in place you can now do this:

```
if(govuk.hasFeatures('query')) {
  // the browser supports govuk.query
  // (which we know under the hood supports querySelectorAll)
  // so now we can enhance the interface using that api
  // this is what progressive enhancement is
}
```

Might be worth reading the articles I cited above in full, as opposed to skimming them. Here's a good one:

http://peter.michaux.ca/articles/feature-detection-state-of-the-art-browser-scripting

**Comment by adamsilver** [comment]
> So I didn't understand how event delegation itself is a concern that influences whether or not you should polyfill.

Because part of the temptation to want to use polyfills stems for the fact that you can just interface directly with host objects. Where as actually, to be able to delegate events (and that's just one of a million examples) you'd need an abstraction anyway.

**Comment by adamsilver** [comment]
> My approach to adopting semi-supported technology on public websites is that if it cannot easily be polyfilled

That's a real unfortunate approach you're taking there. For example, geo location. You can't polyfill that. But with a simple feature detected wrapper (see above etc etc etc) then you can enhance for browsers that support it without hurting users of browsers that don't.

**Comment by pauldwaite** [comment]
@adamsilver 

> If you ignore the work of writing the polyfill, then you ignore the work of writing the alternative to polyfills.

If the work of writing the alternative has already been done by other people, sure, ignore it. My understanding of this proposal is for people working on GOV.UK code to do that work.

**Comment by adamsilver** [comment]
> If the work of writing the alternative has already been done by other people, sure, ignore it. My understanding of this proposal is for people working on GOV.UK code to do that work.

Or to port, and test polyfills perhaps? Or to integrate/consume a service and write guidance around them for GOVUK front end. Etc.

**Comment by joelanman** [comment]
I think its good to acknowledge there would be more work, that has to be part of the discussion, even if it's worth doing the work.

**Comment by nickcolley** [comment]
Hey everyone,

Thanks so much for the conversation, it gives us a good basis to make a decision on.

Will be making a decision on this next sprint. 👍

**Comment by leekowalkowski-hmrc** [comment]
> With the wrapper code in place you can now do this:
> `if(govuk.hasFeatures('query'))`

But you could already do `if(document.querySelectorAll)` which is what I would prefer*, especially when `query` is an arbitrary synonym that would have to be learned, having to discover these associations could be frustrating.

* Unless I decide that `querySelector` is within my 'lowest-common-denominator', which these days, I do.  In the days when I did not, I avoided `querySelector` altogether.

This is not one of the situations the feature detection article warned about (because there is always a document object available, you can do it as early as you need to).

> That's a real unfortunate approach you're taking there. For example, geo location.

Ah, no, sorry, I'm not saying I avoid the technology completely in these situations, we can absolutely embrace features for the users that have it via progressive enhancement.  

What I don't like is when we try to create an equivalent experience for users that lack such features, especially because in my experience as an end user, the 'equivalent' I am given is inferior at best, broken at worst, and just frustrating when I know a simple alternative technique exists that just always works (e.g. enter a date as a single field instead of being forced to interact with a flaky date picker).

> should all developers/designers who are choosing to use / relying on a library have/want/need to learn the native/host APIs?

No, that's the point of using the library.

> if they do want to learn the APIs (and that's a big if imo), can't they just look under the hood as opposed to be forced to be exposed to it?

They could, except many libraries are over-complicated and almost impossible to read.  I would say they're not the easiest source for learning.

> And let's say that all developers should be exposed to the native APIs because “we” think that it's best for them to know. Isn't it best for us to show them how to build a robust, feature-detected library of functions :)

Indeed.  As long as we don't do potentially confusing things like rename 'querySelector' to 'query'.

When I was in the HTML Working Group, I remember we were occasionally reminded about the [priority of constituences](https://www.w3.org/TR/html-design-principles/#priority-of-constituencies).  

Basically the priority is the needs of: users > web developers > browser developers > W3C > technical purity.

The argument that we always avoid polyfills because they manipulate host objects, which is sometimes dangerous, sounds to me like making an argument for technical purity.  Which means we need to consider if that is going to have a knock-on cost for developers or users.

An example of this is BEM vs CSS with an actual cascade, BEM makes a CSS author's life easier, but at the cost of the HTML author, for whom it is a cognitive nightmare.

**Comment by adamsilver** [comment]
> The argument that we always avoid polyfills because they manipulate host objects, which is sometimes dangerous, sounds to me like making an argument for technical purity. Which means we need to consider if that is going to have a knock-on cost for developers or users.

More than valid concern. For me that's not what it is. But again, really valid and good to add to the discussion.

> As long as we don't do potentially confusing things like rename 'querySelector' to 'query'.

Yep, the name isn't ideal. The name could totally be changed.

**Comment by frankieroberto** [comment]
> An example of this is BEM vs CSS with an actual cascade, BEM makes a CSS author's life easier, but at the cost of the HTML author, for whom it is a cognitive nightmare.

Ha, yes, agree with this 100%! All those long abbreviated class names!

**Comment by alex-ju** [comment]
The GOV.UK Design System team had a meeting discussing the proposal and carefully considered all the points raised in this issue.

We have decided to continue with the polyfill approach.

We believe that the community around polyfills is more mature and will help others building components outside of GOV.UK Frontend without a reliance on our team to maintain a utility library.

We accept and understand the risk around the extension of native APIs, given that we are intending to use polyfills that are aiming to be spec compliant.

In cases where polyfills are not yet available or would result in poor behaviour, we will consider ‘cutting the mustard’ (as can be seen in the [Accessible Autocomplete project](https://github.com/alphagov/accessible-autocomplete)) or using utility functions.

Thanks everyone for contributing.
